src/main/java/hu/akarnokd/rxjava2/operators/FlowableWindowPredicate.java

```diff
@@ -0,0 +1,206 @@
+    }
+
+    @Override
+    public Publisher<Flowable<T>> apply(Flowable<T> upstream) {
+        return new FlowableWindowPredicate<T>(upstream, predicate, mode, bufferSize);
+    }
+
+    static final class WindowPredicateSubscriber<T>
+    extends AtomicInteger
+    implements ConditionalSubscriber<T>, Subscription, Runnable {
+
+        final Subscriber<? super Flowable<T>> actual;
+
+        final Predicate<? super T> predicate;
+
+        final Mode mode;
+
+        final int bufferSize;
+
+        final AtomicBoolean cancelOnce;
+
+        Subscription s;
+
+        UnicastProcessor<T> window;
+
+        WindowPredicateSubscriber(Subscriber<? super Flowable<T>> actual,
+                Predicate<? super T> predicate, Mode mode,
+                int bufferSize) {
+            super(1);
+            this.actual = actual;
+            this.predicate = predicate;
+            this.mode = mode;
+            this.bufferSize = bufferSize;
+            this.cancelOnce = new AtomicBoolean();
+        }
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            if (SubscriptionHelper.validate(this.s, s)) {
+                this.s = s;
+                actual.onSubscribe(this);
+            }
+        }
+
+        @Override
+        public void onNext(T t) {
+            if (!tryOnNext(t)) {
+                s.request(1);
+            }
+        }
+
+        @Override
+        public boolean tryOnNext(T t) {
+            UnicastProcessor<T> w = window;
+            if (w == null) {
+                w = newWindow();
+            }
+
+            boolean b;
+
+            try {
+                b = predicate.test(t);
+            } catch (Throwable ex) {
+                Exceptions.throwIfFatal(ex);
+                s.cancel();
+                actual.onError(ex);
+                return true;
+            }
+
+            if (b) {
+                if (mode == Mode.AFTER) {
+                    w.onNext(t);
+                }
+                w = newWindow();
+                if (mode == Mode.BEFORE) {
+                    w.onNext(t);
+                }
+            } else {
+                w.onNext(t);
+            }
+            return b;
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            Processor<T, T> w = window;
+            if (w != null) {
+                window = null;
+                w.onError(t);
+            }
+
+            actual.onError(t);
+        }
+
+        @Override
+        public void onComplete() {
+            Processor<T, T> w = window;
+            if (w != null) {
+                window = null;
+                w.onComplete();
+            }
+
+            actual.onComplete();
+        }
+
+        @Override
+        public void request(long n) {
+            s.request(n);
+        }
+
+        @Override
+        public void cancel() {
+            if (cancelOnce.compareAndSet(false, true)) {
+                s.cancel();
```

Comment:
This will still cancel the upstream immediately. You should call `run()` so the decrement reaching zero should only call `cancel`.